File: src/services/LocationFilter.ts
import type { TrackPoint } from './TrackDataSource';

export type TrackPointRejectReason =
  | 'invalid'
  | 'accuracy'
  | 'too_close'
  | 'too_soon'
  | 'jump';

export type TrackPointFilterProfile = {
  maxAcceptedAccuracyMeters: number;
  minDistanceMeters: number;
  minIntervalMs: number;
  maxSpeedMetersPerSecond: number;
};

export type TrackPointAcceptance =
  | {
      accept: true;
      distanceMeters: number | null;
      elapsedMs: number | null;
    }
  | {
      accept: false;
      reason: TrackPointRejectReason;
      distanceMeters: number | null;
      elapsedMs: number | null;
    };

const EARTH_RADIUS_METERS = 6371008.8;

function toRadians(degrees: number): number {
  return (degrees * Math.PI) / 180;
}

export function distanceMeters(a: TrackPoint, b: TrackPoint): number {
  const lat1 = toRadians(a.latitude);
  const lat2 = toRadians(b.latitude);
  const deltaLat = toRadians(b.latitude - a.latitude);
  const deltaLng = toRadians(b.longitude - a.longitude);

  const sinLat = Math.sin(deltaLat / 2);
  const sinLng = Math.sin(deltaLng / 2);
  const h =
    sinLat * sinLat +
    Math.cos(lat1) * Math.cos(lat2) * sinLng * sinLng;
  const clamped = Math.min(1, Math.max(0, h));

  return (
    2 *
    EARTH_RADIUS_METERS *
    Math.atan2(Math.sqrt(clamped), Math.sqrt(1 - clamped))
  );
}

function hasValidPosition(point: TrackPoint): boolean {
  return (
    Number.isFinite(point.longitude) &&
    Number.isFinite(point.latitude) &&
    Math.abs(point.longitude) <= 180 &&
    Math.abs(point.latitude) <= 90 &&
    Number.isFinite(point.timestamp) &&
    point.timestamp > 0
  );
}

export function shouldAcceptPoint(
  prevAccepted: TrackPoint | null,
  candidate: TrackPoint,
  profile: TrackPointFilterProfile,
): TrackPointAcceptance {
  if (!hasValidPosition(candidate)) {
    return { accept: false, reason: 'invalid', distanceMeters: null, elapsedMs: null };
  }

  if (
    candidate.accuracy !== null &&
    candidate.accuracy !== undefined &&
    (!Number.isFinite(candidate.accuracy) ||
      candidate.accuracy > profile.maxAcceptedAccuracyMeters)
  ) {
    return { accept: false, reason: 'accuracy', distanceMeters: null, elapsedMs: null };
  }

  if (!prevAccepted) {
    return { accept: true, distanceMeters: null, elapsedMs: null };
  }

  const elapsedMs = candidate.timestamp - prevAccepted.timestamp;
  if (!Number.isFinite(elapsedMs) || elapsedMs < 0) {
    return { accept: false, reason: 'invalid', distanceMeters: null, elapsedMs: null };
  }

  const distance = distanceMeters(prevAccepted, candidate);
  if (!Number.isFinite(distance)) {
    return { accept: false, reason: 'invalid', distanceMeters: null, elapsedMs };
  }

  if (elapsedMs < profile.minIntervalMs) {
    return { accept: false, reason: 'too_soon', distanceMeters: distance, elapsedMs };
  }

  if (distance < profile.minDistanceMeters) {
    return { accept: false, reason: 'too_close', distanceMeters: distance, elapsedMs };
  }

  const elapsedSeconds = elapsedMs / 1000;
  if (
    elapsedSeconds > 0 &&
    distance / elapsedSeconds > profile.maxSpeedMetersPerSecond
  ) {
    return { accept: false, reason: 'jump', distanceMeters: distance, elapsedMs };
  }

  return { accept: true, distanceMeters: distance, elapsedMs };
}

File: footprint/Services/LocationFilter.swift
import CoreLocation
import Foundation

enum TrackPointRejectReason: String {
    case invalid
    case accuracy
    case tooClose
    case tooSoon
    case jump
}

enum TrackPointDecision {
    case accept(distanceMeters: Double?, elapsedSeconds: TimeInterval?)
    case reject(reason: TrackPointRejectReason, distanceMeters: Double?, elapsedSeconds: TimeInterval?)
}

nonisolated enum LocationFilter {
    private static let earthRadiusMeters = 6_371_008.8

    static func distanceMeters(from a: TrackPoint, to b: TrackPoint) -> Double {
        let lat1 = a.latitude * .pi / 180
        let lat2 = b.latitude * .pi / 180
        let deltaLat = (b.latitude - a.latitude) * .pi / 180
        let deltaLng = (b.longitude - a.longitude) * .pi / 180

        let sinLat = sin(deltaLat / 2)
        let sinLng = sin(deltaLng / 2)
        let h = sinLat * sinLat + cos(lat1) * cos(lat2) * sinLng * sinLng
        let clamped = min(1, max(0, h))
        return 2 * earthRadiusMeters * atan2(sqrt(clamped), sqrt(1 - clamped))
    }

    static func shouldAccept(previous: TrackPoint?, candidate: TrackPoint, profile: RecordingProfile) -> TrackPointDecision {
        guard candidate.longitude.isFinite,
              candidate.latitude.isFinite,
              abs(candidate.longitude) <= 180,
              abs(candidate.latitude) <= 90,
              candidate.timestampMs > 0
        else {
            return .reject(reason: .invalid, distanceMeters: nil, elapsedSeconds: nil)
        }

        if let accuracy = candidate.accuracy,
           (!accuracy.isFinite || accuracy > profile.maxAcceptedAccuracyMeters) {
            return .reject(reason: .accuracy, distanceMeters: nil, elapsedSeconds: nil)
        }

        guard let previous else {
            return .accept(distanceMeters: nil, elapsedSeconds: nil)
        }

        let elapsedSeconds = TimeInterval(candidate.timestampMs - previous.timestampMs) / 1000
        guard elapsedSeconds >= 0, elapsedSeconds.isFinite else {
            return .reject(reason: .invalid, distanceMeters: nil, elapsedSeconds: nil)
        }

        let distance = distanceMeters(from: previous, to: candidate)
        guard distance.isFinite else {
            return .reject(reason: .invalid, distanceMeters: nil, elapsedSeconds: elapsedSeconds)
        }

        if elapsedSeconds < profile.minIntervalSeconds {
            return .reject(reason: .tooSoon, distanceMeters: distance, elapsedSeconds: elapsedSeconds)
        }

        if distance < profile.minDistanceMeters {
            return .reject(reason: .tooClose, distanceMeters: distance, elapsedSeconds: elapsedSeconds)
        }

        if elapsedSeconds > 0, distance / elapsedSeconds > profile.maxSpeedMetersPerSecond {
            return .reject(reason: .jump, distanceMeters: distance, elapsedSeconds: elapsedSeconds)
        }

        return .accept(distanceMeters: distance, elapsedSeconds: elapsedSeconds)
    }
}
